dhcpv6_relay: initial import of a lightweight DHCPv6 relay agent

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This provides a DHCPv6 relay agent based on `sock_async` which allows for multihop-operation with DHCPv6.

This PR is very much WIP as it is to this point only compile tested, I just wanted to publish it already so people know this is coming ;-).
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
This PR is untested so far. I will add tests as soon as I get to it.
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
None
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented:
I added a scapy-based test application (i.e. it requires root). One-by-one the tests pass, but altogether (or on a non-native board for that matter) they have some issues. Will investigate later. I am also planning to provide a routine to hook this into the DHCPv6 client thread when run on a border router (BR) so the BR can run as a relay agent without an extra thread when the DHCPv6 client is compiled in.

miri64 commented:
> I am also planning to provide a routine to hook this into the DHCPv6 client thread when run on a border router (BR) so the BR can run as a relay agent without an extra thread when the DHCPv6 client is compiled in.

This is currently not possible, since the DHCPv6 client does not use `sock_async`. I.e. it could happen that waiting for a reply from the DHCPv6 server (calling `sock_udp_recv` on the client sock) causes DHCPv6 relay messages to be lost in the meantime.

miri64 commented:
With https://github.com/RIOT-OS/RIOT/pull/16643, this PR's tests should now pass on native.

miri64 commented:
Ok, I refactored the tests to an object-oriented `unittest.TestCase` which allowed me to store most of the state I need for the tests in the object, which reduces the number of UART-operations I need to do with the `samr21-xpro`. As such, now that the "medium" is much cleaner, the tests are able to succeed with `samr21-xpro`, as expected packets are not lost. Now the only thing missing to get this PR out of WIP is the promised border router setup which I will provide ASAP.

miri64 commented:
Updated testing description, no longer WIP.

miri64 commented:
And squashed and rebased to current master.

miri64 commented:
Rebased and squashed to current master to resolve merge conflicts. @benpicco want to take a look?

benpicco commented on sys/net/application_layer/dhcpv6/relay.c:
```diff
@@ -0,0 +1,386 @@
+    res = _join_all_relays_and_server(listen_netif);
+    if (res < 0) {
+        DEBUG("DHCPv6 relay: unable to join All_DHCP_Relay_Agents_and_Servers: "
+
```
```suggestion
```
empty line

benpicco reviewed (commented):
This looks pretty good!

I tested it with the [`setup_taps.sh`](https://github.com/RIOT-OS/RIOT/blob/1834e30ac8b861deb87ed99e4ab84bb1ddef435d/examples/gnrc_networking-subnets/setup_taps.sh) script from #16536 (`examples/gnrc_networking-subnets`) on `native`.

It works quite well with one relay node, but I see some problems if I introduce a second hop / relay node.

<details><summary>changes to the example</summary>

```patch
--- a/examples/gnrc_networking-subnets/Makefile
+++ b/examples/gnrc_networking-subnets/Makefile
@@ -24,13 +24,16 @@ USEMODULE += ps
 USEMODULE += netstats_l2
 USEMODULE += netstats_ipv6
 
+USEMODULE += event_thread
+USEMODULE += dhcpv6_relay
+
 CFLAGS += -DDEBUG_ASSERT_VERBOSE=1
 CFLAGS += -DCONFIG_GNRC_IPV6_NIB_NUMOF=16
 CFLAGS += -DCONFIG_GNRC_IPV6_NIB_RIO_IN_LAST_RA=1
 
 ifeq (native, $(BOARD))
   USEMODULE += netdev_tap
-  CFLAGS += -DCONFIG_GNRC_DHCPV6_CLIENT_6LBR_UPSTREAM=6
+  CFLAGS += -DCONFIG_GNRC_DHCPV6_CLIENT_6LBR_UPSTREAM=7
 else ifeq (1, $(SLIP))
   USEMODULE += slipdev_l2addr
   CFLAGS += -DCONFIG_GNRC_DHCPV6_CLIENT_6LBR_UPSTREAM=5
diff --git a/examples/gnrc_networking/main.c b/examples/gnrc_networking/main.c
index 6301f4291d..3b703cc767 100644
--- a/examples/gnrc_networking/main.c
+++ b/examples/gnrc_networking/main.c
@@ -23,6 +23,10 @@
 #include "shell.h"
 #include "msg.h"
 
+#include "event/thread.h"
+#include "net/dhcpv6/relay.h"
+#include "net/gnrc/netif/internal.h"
+
 #define MAIN_QUEUE_SIZE     (8)
 static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
 
@@ -40,6 +44,17 @@ int main(void)
     msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
     puts("RIOT network stack example application");
 
+    gnrc_netif_t *netif = NULL, *upstream = NULL, *downstream = NULL;
+    while ((netif = gnrc_netif_iter(netif)) != NULL) {
+        if (netif->pid == CONFIG_GNRC_DHCPV6_CLIENT_6LBR_UPSTREAM) {
+            upstream = netif;
+        }
+        else {
+            downstream = netif;
+        }
+    }
+    dhcpv6_relay_init(EVENT_PRIO_LOWEST, downstream->pid, upstream->pid);
+
     /* start shell */
     puts("All up, running the shell now");
     char line_buf[SHELL_DEFAULT_BUFSIZE];
```

</details>

The nodes are then started with

 - `make BOARD=native GATEWAY=1 PORT="tap_a0 tap_b0" all term`
 - `make BOARD=native GATEWAY=1 PORT="tap_b1 tap_c0" term`
 - `make BOARD=native GATEWAY=1 PORT="tap_c1 tap_d0" term`

The first two nodes get their addresses just fine, when the third node is introduced I see dropped messages. (This might be a limitation of `netdev_tap` though)

<details><summary>first node</summary>

```
main(): This is RIOT! (Version: 2021.10-devel-246-g1515c-serial-routing-3)
RIOT network stack example application
All up, running the shell now
> DHCPv6 client: send SOLICIT
DHCPv6 client: resend SOLICIT
DHCPv6 client: received ADVERTISE
DHCPv6 client: scheduling REQUEST
DHCPv6 client: send REQUEST
DHCPv6 client: received REPLY
DHCPv6 client: scheduling RENEW in 10000 sec
DHCPv6 client: scheduling REBIND in 20000 sec
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
gnrc_sock: dropped message to 0x566370d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::c8f4:13ff:fece:3f43%8]:547
```

</details>

<details><summary>second node</summary>

```
RIOT network stack example application
All up, running the shell now
> DHCPv6 client: send SOLICIT
DHCPv6 client: resend SOLICIT
DHCPv6 client: received ADVERTISE
DHCPv6 client: scheduling REQUEST
DHCPv6 client: send REQUEST
DHCPv6 client: discarding 81 stale bytes
DHCPv6 client: discarding 81 stale bytes
DHCPv6 client: discarding 81 stale bytes
DHCPv6 client: received REPLY
DHCPv6 client: scheduling RENEW in 10000 sec
DHCPv6 client: scheduling REBIND in 20000 sec
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [::%8]:546
DHCPv6 relay: forwarding reply towards target failed: 22
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
gnrc_sock: dropped message to 0x565d40d0 (was full)
gnrc_sock: dropped message to 0x565d41b0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
gnrc_sock: dropped message to 0x565d41b0 (was full)
gnrc_sock: dropped message to 0x565d40d0 (was full)
gnrc_sock: dropped message to 0x565d41b0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
gnrc_sock: dropped message to 0x565d40d0 (was full)
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
DHCPv6 relay: forwarding reply towards target [fe80::3c27:6dff:fe25:e95d%8]:546
```

</details>

<details><summary>third node</summary>

```
RIOT network stack example application
All up, running the shell now
> DHCPv6 client: send SOLICIT
DHCPv6 client: resend SOLICIT
DHCPv6 client: received ADVERTISE
DHCPv6 client: scheduling REQUEST
DHCPv6 client: send REQUEST
DHCPv6 client: discarding 81 stale bytes
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
DHCPv6 client: resend REQUEST
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
gnrc_sock: dropped message to 0x565acbb0 (was full)
```

</details>

miri64 commented on sys/net/application_layer/dhcpv6/relay.c:
```diff
@@ -0,0 +1,386 @@
+            }
+        }
+        else {
+            LOG_WARNING("DHCPv6 relay: auto init failed, more than 1 interface\n");
```
It isn't this is why this warning is here: if there is more than one interface, there is no automated way to determine which interface is upstream (towards the server) and which is downstream (towards the client). So the user needs to manually bootstrap.

miri64 replied:
Sure I could just assume the topology, based on some network interface parameters, but I rather would not.

miri64 commented:
> The first two nodes get their addresses just fine, when the third node is introduced I see dropped messages. (This might be a limitation of `netdev_tap` though)

You did not the mention in the DHCPv6 RFC about PD and relay nodes, right? Or did you use IA_NA from #16228 to "get their addresses"?

miri64 commented:
> You did not the mention in the DHCPv6 RFC about PD and relay nodes, right?

See [RFC 8415, section 19.1.3](https://datatracker.ietf.org/doc/html/rfc8415#section-19.1.3):

>   A relay agent forwards messages containing prefix delegation options
>   in the same way as it would relay addresses (i.e., per
>   Sections 19.1.1 and 19.1.2).
>
>   If a server communicates with a client through a relay agent about
>   delegated prefixes, the server may need a protocol or other
>   out-of-band communication to configure routing information for
>   delegated prefixes on any router through which the client may forward
>   traffic.

miri64 commented:
IT could however also be that there is simply a circle happening. See

https://github.com/RIOT-OS/RIOT/blob/8069fcbff4b2bb01a5cefd3842a0890d3972c502/sys/net/application_layer/dhcpv6/relay.c#L260-L261

benpicco commented:
Yes I was using DHCPv6 PD to get the prefix on the downstream interface, the upstream one got it's prefix via Router Advertisements of the next hop node:

#### second node
```
ifconfig
Iface  8  HWaddr: 52:75:35:FA:30:BC 
          L2-PDU:1500  MTU:1500  HL:64  RTR  
          RTR_ADV  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::5075:35ff:fefa:30bc  scope: link  VAL
          inet6 addr: 2001:db8:0:70:5075:35ff:fefa:30bc  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:fffa:30bc
          inet6 group: ff02::1:2
          
          Statistics for Layer 2
            RX packets 18  bytes 2008
            TX packets 27 (Multicast: 6)  bytes 3575
            TX succeeded 27 errors 0
          Statistics for IPv6
            RX packets 18  bytes 1756
            TX packets 27 (Multicast: 6)  bytes 3197
            TX succeeded 27 errors 0

Iface  7  HWaddr: CA:F4:13:CE:3F:43 
          L2-PDU:1500  MTU:1500  HL:64  RTR  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::c8f4:13ff:fece:3f43  scope: link  VAL
          inet6 addr: 2001:db8:0:20:c8f4:13ff:fece:3f43  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ffce:3f43
          
          Statistics for Layer 2
            RX packets 42  bytes 6709
            TX packets 38 (Multicast: 34)  bytes 4932
            TX succeeded 38 errors 0
          Statistics for IPv6
            RX packets 42  bytes 6121
            TX packets 38 (Multicast: 34)  bytes 4400
            TX succeeded 38 errors 0

> nib prefix
nib prefix
2001:db8:0:20::/60 dev #7  expires 39847 sec deprecates 29847 sec
2001:db8:0:70::/60 dev #8  expires 39848 sec deprecates 29848 sec
```

miri64 commented:
> IT could however also be that there is simply a circle happening. See
> 
> https://github.com/RIOT-OS/RIOT/blob/8069fcbff4b2bb01a5cefd3842a0890d3972c502/sys/net/application_layer/dhcpv6/relay.c#L260-L261

Ah no, since you are using multiple interfaces this is highly unlikely.

miri64 commented:
> gnrc_sock: dropped message to 0x565acbb0 (was full)
> gnrc_sock: dropped message to 0x565acbb0 (was full)
> gnrc_sock: dropped message to 0x565acbb0 (was full)
> gnrc_sock: dropped message to 0x565acbb0 (was full)
> gnrc_sock: dropped message to 0x565acbb0 (was full)

Mhh not sure what this is about. This looks more like the potential trouble maker. Will investigate.

miri64 commented:
Which address belongs to which node?

miri64 commented:
> DHCPv6 relay: forwarding reply towards target [::%8]:546
> DHCPv6 relay: forwarding reply towards target failed: 22

These also look concerning. The peer-address should not be unspecified.

miri64 commented:
Can you maybe provide a wireshark dump of these communications?

miri64 commented:
At least the unspecified address issue should now be dealt with with the latest suite of commits.

benpicco commented:
There you go: [br0.pcapng.zip](https://github.com/RIOT-OS/RIOT/files/6902824/br0.pcapng.zip)

Not sure which of the bridges is the most interesting, but if you want to reproduce this on your own machine, you only need this patch

<details><summary>patch to gnrc_networking</summary>

```patch
diff --git a/examples/gnrc_networking/Makefile b/examples/gnrc_networking/Makefile
index a59b776599..6a8c5f44c3 100644
--- a/examples/gnrc_networking/Makefile
+++ b/examples/gnrc_networking/Makefile
@@ -32,6 +32,13 @@ USEMODULE += netstats_l2
 USEMODULE += netstats_ipv6
 USEMODULE += netstats_rpl
 
+USEMODULE += event_thread
+USEMODULE += dhcpv6_relay
+
+USEMODULE += gnrc_dhcpv6_client_simple_pd
+CFLAGS += -DCONFIG_GNRC_DHCPV6_CLIENT_6LBR_UPSTREAM=7
+CFLAGS += -DNETDEV_TAP_MAX=2
+
 # Optionally include DNS support. This includes resolution of names at an
 # upstream DNS server and the handling of RDNSS options in Router Advertisements
 # to auto-configure that upstream DNS server.
diff --git a/examples/gnrc_networking/main.c b/examples/gnrc_networking/main.c
index 6301f4291d..3b703cc767 100644
--- a/examples/gnrc_networking/main.c
+++ b/examples/gnrc_networking/main.c
@@ -23,6 +23,10 @@
 #include "shell.h"
 #include "msg.h"
 
+#include "event/thread.h"
+#include "net/dhcpv6/relay.h"
+#include "net/gnrc/netif/internal.h"
+
 #define MAIN_QUEUE_SIZE     (8)
 static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
 
@@ -40,6 +44,17 @@ int main(void)
     msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
     puts("RIOT network stack example application");
 
+    gnrc_netif_t *netif = NULL, *upstream = NULL, *downstream = NULL;
+    while ((netif = gnrc_netif_iter(netif)) != NULL) {
+        if (netif->pid == CONFIG_GNRC_DHCPV6_CLIENT_6LBR_UPSTREAM) {
+            upstream = netif;
+        }
+        else {
+            downstream = netif;
+        }
+    }
+    dhcpv6_relay_init(EVENT_PRIO_LOWEST, downstream->pid, upstream->pid);
+
     /* start shell */
     puts("All up, running the shell now");
     char line_buf[SHELL_DEFAULT_BUFSIZE];
``` 
</details>

and this script
<details><summary>setup.sh</summary>

```bash
#!/bin/bash

PREFIX_UPLINK=2001:dbc::/64
PREFIX_DOWNLINK=2001:db8::/32

RIOTBASE=../..
RIOTTOOLS=$RIOTBASE/dist/tools

# add address to loopback so we can test reaching an
# address outside the test net
sudo ip a a fd00:dead:beef::1/128 dev lo

# create topology
echo "create new interfaces"
sudo $RIOTTOOLS/tapsetup/tapsetup -b br0 -t tap_a -c 1
sudo $RIOTTOOLS/tapsetup/tapsetup -b br1 -t tap_b -c 3
sudo $RIOTTOOLS/tapsetup/tapsetup -b br2 -t tap_c -c 3
sudo $RIOTTOOLS/tapsetup/tapsetup -b br3 -t tap_d -c 1

# start radvd & dhcpd
echo "starting daemons"
sudo $RIOTTOOLS/radvd/radvd.sh -c br0 $PREFIX_UPLINK
sudo mkdir /run/kea

# we must start kea *after* connecting to the tap device
echo please run
echo sudo $RIOTTOOLS/dhcpv6-pd_ia/dhcpv6-pd_ia.py -g 60 br0 $PREFIX_DOWNLINK
```
</details>

Now just run

 - `make PORT="tap_a0 tap_b0" all  term`
 - `make PORT="tap_b1 tap_c0" all  term`
 - `make PORT="tap_c1 tap_d0" all  term`

miri64 commented:
> setup.sh

Not 100% sure but I think you are trying what [veth(4)](https://man7.org/linux/man-pages/man4/veth.4.html) is doing with bridges :laughing:

miri64 commented:
> There you go: [br0.pcapng.zip](https://github.com/RIOT-OS/RIOT/files/6902824/br0.pcapng.zip)

Ah, you seem to have a similar problem like I had the other day (with libCoAP in my case however): For some reason newer Linux kernels decided to rely on some Ethernet offloading which screws up the UDP checksum calculation. This seems to be a reoccuring issue, as using [this workaround](https://github.com/raspberrypi/linux/issues/1944#issuecomment-290982933) seemed to work for me. So for all bridges and TAP interfaces I did

```console
$ sudo ethtool --offload "<iface>" rx off tx off
```

miri64 commented:
Will try to reproduce however ASAP.

miri64 commented:
> > There you go: [br0.pcapng.zip](https://github.com/RIOT-OS/RIOT/files/6902824/br0.pcapng.zip)
> 
> Ah, you seem to have a similar problem like I had the other day (with libCoAP in my case however): For some reason newer Linux kernels decided to rely on some Ethernet offloading which screws up the UDP checksum calculation ([edit] which causes incoming UDP packets to be dropped by RIOT). This seems to be a reoccuring issue, as using [this workaround](https://github.com/raspberrypi/linux/issues/1944#issuecomment-290982933) seemed to work for me. So for all bridges and TAP interfaces I did
> 
> ```
> $ sudo ethtool --offload "<iface>" rx off tx off
> ```

Yepp, with

```sh
ip link | grep -o -e 'br[0-9]\+' -e 'tap_[abc][0-9]\+' |  xargs -I{} sudo ethtool --offload {} rx off tx off
```

being run before I start the DHCPv6 server and then rebooting all nodes (starting with the most upstream one and then going down the tree), I get the downlink prefix for all running `native` instances.

miri64 commented:
> […] and then rebooting all nodes (starting with the most upstream one and then going down the tree), I get the downlink prefix for all running `native` instances.

(In case rebooting does not work, restarting the whole `native` process might get better results ;-)).

miri64 commented:
> ```shell
> sudo $RIOTTOOLS/tapsetup/tapsetup -b br1 -t tap_b -c 3
> sudo $RIOTTOOLS/tapsetup/tapsetup -b br2 -t tap_c -c 3
> ```

Why three btw when I only use two of each bridge?

benpicco commented:
> Why three btw when I only use two of each bridge?

That was just copied from my subnet test where I also have leaf nodes on each level

miri64 commented:
> Yepp, with
> 
> ```shell
> ip link | grep -o -e 'br[0-9]\+' -e 'tap_[abc][0-9]\+' |  xargs -I{} sudo ethtool --offload {} rx off tx off
> ```
> 
> being run before I start the DHCPv6 server and then rebooting all nodes (starting with the most upstream one and then going down the tree), I get the downlink prefix for all running `native` instances.

(as I only get output for the bridges with that snippet, I think setting offloading only for the bridges might suffice:

```sh
ip link | grep -e 'br[0-9]\+' |  xargs -I{} sudo ethtool --offload {} rx off tx off
```

benpicco commented:
Hm I think this is a red herring. The UDP packets are dropped because the mbox in `gnrc_sock` is full and stays full. E.g. when the 1st node in the tree is in the `gnrc_sock: dropped message to 0x565b90d0 (was full)` state and I reboot the second node (wich previously got a DHCPv6 lease) the first node will still print `gnrc_sock: dropped message to 0x565b90d0 (was full)`.

I tried with offloading disabled and I still see the same issue.

miri64 commented:
Then I am afraid I can't reproduce your issue locally :-/.

benpicco commented:
How many messages can this `mbox` store?

benpicco commented:
When I run this on hardware (`same54-xpro` with Ethernet uplink, WeAct-F411CE wired up over UART / `slipdev_l2`) I can only observe the SOLICIT being relayed by the `same54-xpro`, it gets no reply.

Maybe that's a limitation of my router though. It answers to `ff02::1:2` as it should, I also subscribed to that address so I can capture the DHCPv6 request: [dhcpv6-repay.pcapng.gz](https://github.com/RIOT-OS/RIOT/files/6945021/dhcpv6-repay.pcapng.gz)

miri64 commented:
So how do we continue? I am unable to reproduce your errors (either because the error does not occur for me or because I do not have the hardware).

benpicco commented:
Well the problem with the multiple relay hops appears to be caused by a lower layer, the code from this PR looks fine.
I would have hoped we could get closer to the cause of this (maybe just a queue that's too small, but then again packets appear to be send multiple times more than expected).

The single-hop relay case works fine though (with KEA), so if you don't want to investigate this further, feel free to squash.

miri64 commented:
> […] so if you don't want to investigate this further, feel free to squash.

In the longtime I'd like to investigate further, if I am able to reproduce this, but for now I just would prefer to have this in (also with #16228 being merged soon and multi-hop in mind).

miri64 commented:
Squashed

miri64 commented:
Updated `Makefile.ci` of the test. Also rebased to resolve merge conflict with current master.

miri64 commented:
And fixed the broken `Makefile.ci` :roll_eyes:

miri64 commented:
I give #16228 precedence over merging this PR... its the older PR and there _will_ be merge conflicts :-).

miri64 commented:
Ok, there need to be some fixes there anyway, so go.
